odl/deform/linearized.py

```diff
@@ -0,0 +1,342 @@
+        [0.0, 0.0, 1.0, 0.5, 0.0]
+        """
+        if domain is None:
+            if not isinstance(template, DiscreteLpVector):
+                raise TypeError('`template` must be a `DiscreteLpVector`'
+                                'instance if `domain` is None, got {!r}'
+                                ''.format(template))
+
+            domain = template.space.vector_field_space
+        else:
+            if not isinstance(domain, ProductSpace):
+                raise TypeError('`domain` {!r} not a `ProductSpace`'
+                                ''.format(domain))
+            if not domain.is_power_space:
+                raise TypeError('`domain` {!r} not a product'
+                                'space'.format(domain))
+            if not isinstance(domain[0], DiscreteLp):
+                raise TypeError('`domain[0]` {!r} not a `DiscreteLp`'
+                                ''.format(domain))
+            if not domain[0].is_rn:
+                raise TypeError('`domain[0]` {!r} not a real space'
+                                ''.format(domain[0]))
+
+            template = domain[0].element(template)
+
+        Operator.__init__(self, domain, template.space, linear=False)
+
+        self._template = template
+
+    def _call(self, displacement, out=None):
+        """Implementation of ``self(displacement)``.
+
+        Parameters
+        ----------
+        displacement : `domain` element
+            The displacement field used in the linearized deformation.
+        """
+        return _linear_deform(self._template, displacement, out)
+
+    def derivative(self, displacement):
+        """Derivative of the operator in ``displacement``.
+
+        Parameters
+        ----------
+        displacement : `domain` element-like
+            Point that the derivative need to be computed at.
+
+        Returns
+        -------
+        derivative : `PointwiseInner`
+            The derivative evaluated at ``displacement``.
+        """
+        # To implement the complex case we need to be able to embed the real
+        # vector field space into the range of the gradient. Issue #59.
+        if not self.range.is_rn:
+            raise NotImplementedError('derivative not implemented for complex '
+                                      'spaces.')
+
+        displacement = self.domain.element(displacement)
+
+        # TODO allow users to select what method to use here.
+        grad = Gradient(domain=self.range, method='central',
+                        padding_method='symmetric')
+        grad_templ = grad(self._template)
+        def_grad = self.domain.element(
+            [_linear_deform(gf, displacement) for gf in grad_templ])
+
+        return PointwiseInner(self.domain, def_grad)
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        if self.domain == self._template.space.vector_field_space:
+            domain_repr = ''
+        else:
+            domain_repr = ', domain={!r}'.format(self.domain)
```

Comment:
Will this case ever occur? If `domain` is not given during init, we are always in the first case. If it is given, we know that `template in domain[0]` by construction. So when would the condition in the `if` part not hold?